forest/shared/src/cli/mod.rs

```diff
@@ -171,60 +169,112 @@ impl CliOpts {
+        let path = dir.config_dir().join("config.toml");
+        if path.exists() {
+            return Some(ConfigPath::Project(path));
         }
     }
-
-    warn!("No configurations found, using defaults.");
-
     None
 }
 
-fn read_config_or_none(path: PathBuf) -> Option<Config> {
-    let toml = match read_file_to_string(&path) {
-        Ok(t) => t,
-        Err(e) => {
-            warn!("An error occured while reading configuration file at {}. Resorting to default configuration. Error was {}", path.display(), e);
-            return None;
+fn get_keys_flatten(value: toml::Value, keys: &mut HashSet<String>) {
+    match value {
+        toml::Value::Table(map) => {
+            for (k, v) in map.into_iter() {
+                keys.insert(k);
+                get_keys_flatten(v, keys);
+            }
         }
-    };
-
-    match read_toml(&toml) {
-        Ok(cfg) => Some(cfg),
-        Err(e) => {
-            warn!(
-                "Error reading configuration file, opting to defaults. Error was {} ",
-                e
-            );
-            None
+        toml::Value::Array(vec) => {
+            for v in vec.into_iter() {
+                get_keys_flatten(v, keys);
+            }
         }
+        _ => (),
+    }
+}
+
+pub fn warn_for_unknown_keys(path: &Path, config: &Config) {
+    // `config` has been loaded successfully from toml file in `path` so we can always serialize
+    // it back to a valid TOML value or get the TOML value from `path`
+    let file = read_file_to_string(path).unwrap();
+    let value = file.parse::<toml::Value>().unwrap();
+    let mut keys = HashSet::new();
+    get_keys_flatten(value, &mut keys);
+
+    let config_file = toml::to_string(config).unwrap();
+    let config_value = config_file.parse::<toml::Value>().unwrap();
+    let mut config_keys = HashSet::new();
+    get_keys_flatten(config_value, &mut config_keys);
+
+    for k in keys.difference(&config_keys) {
+        warn!("Unknown key \"{k}\" in {}, ignoring", path.display());
     }
 }
 
 /// Print an error message and exit the program with an error code
 /// Used for handling high level errors such as invalid parameters
 pub fn cli_error_and_die(msg: impl AsRef<str>, code: i32) -> ! {
-    error!("Error: {}", msg.as_ref());
+    error!("{}", msg.as_ref());
     std::process::exit(code);
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn get_keys_flatten_must_work() {
+        let value = toml::from_str(
+            r#"
+            folklore = true
+            [myth]
+            author = 'H. P. Lovecraft'
+            entities = [
+                { name = 'Cthulhu' },
+                { name = 'Azathoth' },
+                { name = 'Dagon' },
+            ]
+        "#,
+        )
+        .unwrap();
+
+        let mut keys = HashSet::new();
+        get_keys_flatten(value, &mut keys);
+
+        let expected = HashSet::from_iter(
+            ["folklore", "myth", "author", "entities", "name"]
```

Comment:
Hm, is `myth` a key? Or in other terms, will the `warn_for_unknown_keys` warn about config such as:
```
[data_dir]
client = "blabla"
```
(section trying to be a key)

What I have in mind, is the keys should be "paired" with their respective sections, so:
```
[client]
data_dir = "blabla"
```
would be a single key `client.data_dir`. This would also avoid keys in invalid sections, i.e. 
```
[client]
write_buffer_size = 42 # this should be in `rocks_db` section
```

I'm not sure it makes sense or perhaps you already covered this?

Reply:
Yeah, this poor-man's diff is not 100% accurate as you can see. I can make it better.